File: Source/NeoDelphiShared/Lib/NeoAi/Entity/RepPropertyValue.pas
unit RepPropertyValue;

interface

uses
  RepObjectBase, TypesConsts, Entity;

type
  TRepPropertyValue = class(TRepObjectBase)
  public
    type
      TLinkType = (ltNone, ltEntity, ltAttrKey, ltEventKey);
      TOperatorType = (otNone, otEquals, otSmaller, otSmallerOrEqual, otGreater, otGreaterOrEqual, otDiffers);
      TOperatorTypes = set of TOperatorType;
  private
    FLinkObject: TEntity;
    FValue: string;
    FRepPropertyKey: TEntity;
    FKeyId: TId;
    FOperatorType: TOperatorType;
    FTargetValueId: TId;
    FTargetEntityId: TId;
    FTargetKeyId: TId;
  public
    constructor Create(ARepPropertyKey: TEntity; AnOperatorType: TOperatorType; const AValue: string); overload;
    constructor Create(ARepPropertyKey: TEntity; AnOperatorType: TOperatorType; ALinkType: TRepPropertyValue.TLinkType; ALinkObject: TEntity); overload;

    property LinkObject: TEntity read FLinkObject write FLinkObject;
    property RepPropertyKey: TEntity read FRepPropertyKey write FRepPropertyKey;
  published
    property Id;
    property Kids;
    property TargetEntityId: TId read FTargetEntityId write FTargetEntityId;
    property TargetKeyId: TId read FTargetKeyId write FTargetKeyId;
    property TargetValueId: TId read FTargetValueId write FTargetValueId;
    property OperatorType: TOperatorType read FOperatorType write FOperatorType;
    property KeyId: TId read FKeyId write FKeyId;
    property Value: string read FValue write FValue;
  end;

implementation

{ TRepPropertyValue }

constructor TRepPropertyValue.Create(ARepPropertyKey: TEntity; AnOperatorType: TOperatorType; const AValue: string);
begin
  Create;
  FOperatorType := AnOperatorType;
  FRepPropertyKey := ARepPropertyKey;
  FValue := AValue;
end;

constructor TRepPropertyValue.Create(ARepPropertyKey: TEntity; AnOperatorType: TOperatorType;
  ALinkType: TRepPropertyValue.TLinkType; ALinkObject: TEntity);
begin
  Create;
  case ALinkType of
    ltEntity:
      FTargetEntityId := ALinkObject.Id;
    ltAttrKey:
      FTargetKeyId := ALinkObject.Id;
    ltEventKey:
      FTargetValueId := ALinkObject.Id;
  end;
  FLinkObject := ALinkObject;
  FOperatorType := AnOperatorType;
  FRepPropertyKey := ARepPropertyKey;
end;

initialization
  TRepPropertyValue.RegisterEntityClass;

end.

File: Source/NeoXplora.com/js/classes/rep/RepPropertyValue.js
var TRepPropertyValue_Implementation = {
  
  construct: function(Parent, Id, TargetEntityId, TargetKeyId, TargetValueId, OperatorType, KeyId, Value) {
    this.base(this);
    this.Kids = new NeoX.TStringList();
    this.setParentObject(Parent);
    this.setId(Id);
    this.setTargetEntityId(TargetEntityId);
    this.setTargetKeyId(TargetKeyId);
    this.setTargetValueId(TargetValueId);
    this.setOperatorType(OperatorType);
    this.setKeyId(KeyId);
    this.setValue(Value);
  },
  
  properties: {
    Id: null, //int
    ParentObject: null, //RepPropertyKey
    Kids: null, //array of string : repPropertyKey
    TargetEntityId: null, //int
    TargetKeyId: null, //int
    TargetValueId: null, //int
    OperatorType: null, //(otNone, otEquals, otSmaller, otSmallerOrEqual, otGreater, otGreaterOrEqual, otDiffers)
    KeyId: null, //int
    Value: null, //string
    LinkObject: null //LinkType: (ltNone, ltEntity, ltAttrKey, ltEventKey)
  },
  
  methods: {
    
    addKid: function(key, repPropertyValue) {
      this.Kids.add(key, repPropertyValue);
    }
    
  }

};

Sky.Class.Define("NeoX.TRepPropertyValue", TRepPropertyValue_Implementation);